model_analyzer/result/constraint_manager.py

```diff
@@ -75,3 +81,36 @@ def satisfies_constraints(constraints, run_config_measurement):
                                 return False
 
         return True
+
+    @staticmethod
+    def constraint_failure_percentage(
+            constraints: List[Dict[str, Dict[str, int]]],
+            run_config_measurement: 'RunConfigMeasurement') -> float:
+        """
+        Additive percentage, for every measurement, in every model, of how much 
+        the RCM is failing the constraints by
+        
+        Returns
+        -------
+        Float
+        """
+        failure_percentage = 0
+
+        if constraints:
+            for (i, model_metrics) in enumerate(run_config_measurement.data()):
+                for metric in model_metrics:
+                    if constraints[i] is not None and metric.tag in constraints[
+                            i]:
+                        constraint = constraints[i][metric.tag]
+                        if 'min' in constraint:
+                            if metric.value() < constraint['min']:
```

Comment:
6 levels of nesting -> smells like it needs a refactor

Reply:
maybe something like `_get_failure_percentage(metric, constraint)`, which can also be reused in satisfies_constraints()?

Reply:
I've refactored, but now running into a type check failure I can't solve:

`../../model_analyzer/result/constraint_manager.py:79: error: Argument 2 to "_get_failure_percentage" of "ConstraintManager" has incompatible type "Optional[Dict[str, int]]"; expected "Dict[str, int]"
`

The issue is that `_metric_matches_constraint` can return either: (True, constraint) or (False, None). Any suggestions about how I can get around this?

Reply:
It is pointing out a legit issue. If _metric_matches_constraint() can return <False, None>, then the downstream _get_failure_percentage(metric, constraint) needs to be able to handle "None" for the constraint input arg. I understand that will never actually happen the way the code is structured, but it should still be cleaned up. 

Why do you have to return a tuple from _metric_matches_constraint? Just return a bool, and then if that is true the caller can extract the constraint.

```
if (ConstraintManager._metric_matches_constraint(metric, constraints[i])):
  constraint = constraints[i][metric.tag]
  failure_percentage += ...
```

Reply:
Understood. I've refactored.